include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -320,149 +320,68 @@ namespace internal
-      (void)constraints_in;
-      if (locally_relevant_constraints.empty())
-        return;
-
-      for (ConstraintLine &entry : locally_relevant_constraints)
-        std::sort(entry.entries.begin(),
-                  entry.entries.end(),
-                  [](const auto &l1, const auto &l2) {
-                    return l1.first < l2.first;
+    // First define a helper function that sorts and normalizes the constraints
+    // provided through the function's argument:
+    const auto sort_and_make_unique =
+      [](std::vector<ConstraintLine> &locally_relevant_constraints) {
+        if (locally_relevant_constraints.empty())
+          return;
+
+        // First sort the array of constraints by their index:
+        std::sort(locally_relevant_constraints.begin(),
+                  locally_relevant_constraints.end(),
+                  [](const ConstraintLine &l1, const ConstraintLine &l2) {
+                    return l1.index < l2.index;
                   });
 
-      std::sort(locally_relevant_constraints.begin(),
-                locally_relevant_constraints.end(),
-                [](const ConstraintLine &l1, const ConstraintLine &l2) {
-                  return l1.index < l2.index;
-                });
-
-      const auto equal_with_tol = [](const number d, const number e) {
-        if (std::abs(std::real(d - e)) <
-            std::sqrt(
-              std::numeric_limits<
-                typename numbers::NumberTraits<number>::real_type>::epsilon()))
-          return true;
-        else
-          return false;
+        // It is possible that two processes have computed constraints
+        // for the same DoF differently (for example, in parallel because
+        // they have visited different faces of the same cell because
+        // on different processes, different neighbor cells are artificial.
+        //
+        // This is ok, as long as after resolution of all chains of
+        // constraints we end up with the same constraint. But at this
+        // point, we just don't know yet. We deal with this by simply
+        // dropping constraints for DoFs for which there is a previous
+        // constraint in the list:
+        locally_relevant_constraints.erase(
+          std::unique(locally_relevant_constraints.begin(),
+                      locally_relevant_constraints.end(),
+                      [](const ConstraintLine &a, const ConstraintLine &b) {
+                        return (a.index == b.index);
+                      }),
+          locally_relevant_constraints.end());
+
+        // For those constraints that survive, sort the right hand side arrays:
+        for (ConstraintLine &entry : locally_relevant_constraints)
+          std::sort(entry.entries.begin(),
+                    entry.entries.end(),
+                    [](const auto &l1, const auto &l2) {
+                      return l1.first < l2.first;
+                    });
       };
```

Comment:
This is the original helper function of release 9.5 before #14905 was added.
https://github.com/dealii/dealii/blob/b7e4db145833d6f74502731fe630296b2a1d910a/include/deal.II/lac/affine_constraints.templates.h#L221-L229

Comparing the algorithm (and ignoring your clean-ups), sorting the right hand side arrays in the helper function was the only change you made.

So we are basically reverting #14905. Do you agree @bergbauer @kronbichler?

Reply:
Our use case does not work with this patch but it is also broken on master. It used to work with the originally proposed patch together with the old master branch for which it was written. I don't think one can accomplish setting up constraints consistently in parallel for aggregated finite element methods without merging (at least for continuous finite elements). With the test from https://github.com/dealii/dealii/pull/14905 @kronbichler and I tried to imitate what was happening but it seems we have failed at that.

What do you think about the following plan: I open a new issue for setting up constraints consistently for aggregated finite elements where I try to explain the situation and post a test case. In the meantime, we can merge this if it fixes your current problems.

Reply:
@bergbauer  Yes, please do that. Let's first precisely define the semantics you want, and then we can think about how to implement this. I often find it useful to first write the documentation/specification, and then implement what we say there.